extensions/pkg/webhook/controlplane/genericmutator/mutator.go

```diff
@@ -228,16 +230,42 @@ func findFileWithPath(osc *extensionsv1alpha1.OperatingSystemConfig, path string
 }
 
 func (m *mutator) mutateOperatingSystemConfig(ctx context.Context, gctx gcontext.GardenContext, osc, oldOSC *extensionsv1alpha1.OperatingSystemConfig) error {
+	cluster, err := gctx.GetCluster(ctx)
+	if err != nil {
+		return err
+	}
+
+	kubeletVersion, err := semver.NewVersion(cluster.Shoot.Spec.Kubernetes.Version)
+	if err != nil {
+		return err
+	}
+
 	// Mutate kubelet.service unit, if present
 	if content := getKubeletService(osc); content != nil {
-		if err := m.ensureKubeletServiceUnitContent(ctx, gctx, content, getKubeletService(oldOSC)); err != nil {
+		// Deserialize unit options
+		opts, err := m.unitSerializer.Deserialize(*content)
+		if err != nil {
+			return fmt.Errorf("could not deserialize kubelet.service unit content: %w", err)
+		}
+
+		// Parse kubelet version from unit description if present
+		if opt := extensionswebhook.UnitOptionWithSectionAndName(opts, "Unit", "Description"); opt != nil {
+			if split := strings.Split(opt.Value, " "); len(split) == 3 {
+				kubeletVersion, err = semver.NewVersion(split[2])
+				if err != nil {
+					return err
+				}
+			}
+		}
```

Comment:
IIUC, here we want to fetch the kubelet version of the corresponding worker pool? If yes, parsing of the unit description to do this seems to be odd - obviously it is fragile (easy to regress just but changing the unit description). Wouldn't it be possible from the OSC to fetch the corresponding worker pool name and from the cluster obj of fetch the kubelet version for this worker pool?

Reply:
A similar discussion already happened in https://github.com/gardener/gardener/pull/5256/files/02cf105d9e6d1cf55662f81df0f0949862ce3f9f..8dc4d53618a7c789f8eec7239cd37204ecdd4fb5#diff-48763d885c4bfb36d75cd6ec08744f2408305f1ebdfb81fb54da0025132c908b. Here, the proposal was to introduce a field for the kubelet version in the `OperatingSystemConfig`.

Your proposal is a bit different and similarly implementable, of course. However, I don't see how this would be any better compared to the current proposal: In order to fetch the effective version from the `Cluster` object we need to know the name of the worker pool. Just by looking at the `OperatingSystemConfig` we don't know it. We would have to similarly parse the name of the resource (`cloud-config-<poolname>-<hash>-<os-name>-{original,downloader}`) before looping over `cluster.Shoot.Spec.Provider.Workers[]`. I don't think such approach addresses the concern you raised ("obviously it is fragile (easy to regress just but changing the unit description)").

As an alternative, we could label or annotate the `OperatingSystemConfig` with the name of the pool it got created for and then read this field here in the generic mutator.

Reply:
Yes, I came to the same conclusion that the OSC does not contain in a proper way the pool name and we better introduce a label to the OSC that contains the pool name. I can create an separate issue to improve this handling as discussed here, if you are OK with it.

Reply:
OK, great, I've made the adaptations (see commits 5677076, ebcf638 and ee62fb3).
Probably @timebertt will also like this approach better :)

Reply:
Yes, I definitely like the approach! Thanks to you both :)